### PR TITLE
implement cms api

### PR DESCRIPTION
**Pull request by mo271**

### CODE REVIEW

File: lib/jxl/render_pipeline/stage_from_linear.cc [hunk 1]
```diff
@@ -4,6 +4,8 @@
 // license that can be found in the LICENSE file.
 
 #include "lib/jxl/render_pipeline/stage_from_linear.h"
+#include <cstdio>
+#include "lib/jxl/color_encoding_internal.h"
```
Comment (mo271) [inline lib/jxl/render_pipeline/stage_from_linear.cc:8]:
```suggestion
```

File: lib/jxl/dec_cache.cc [hunk 2]
```diff
@@ -195,23 +196,57 @@ Status PassesDecoderState::PreparePipeline(ImageBundle* decoded,
...
+          output_encoding_info.orig_color_encoding.IsGray());
+    } else {
+      linear_color_encoding = output_encoding_info.linear_color_encoding;
+    }
+
     auto tone_mapping_stage = GetToneMappingStage(output_encoding_info);
     if (tone_mapping_stage) {
       if (!linear) {
         auto to_linear_stage = GetToLinearStage(output_encoding_info);
         if (!to_linear_stage) {
```
Comment (sboukortt) [inline lib/jxl/dec_cache.cc:211]: I think maybe I would swap this condition around (`if (!cond) { a } else { b }` -> `if (cond) { b } else { a }`).

File: lib/jxl/dec_xyb.cc [hunk 3]
```diff
@@ -250,6 +251,8 @@ Status OutputEncodingInfo::MaybeSetColorEncoding(
 
 Status OutputEncodingInfo::SetColorEncoding(const ColorEncoding& c_desired) {
   color_encoding = c_desired;
+  linear_color_encoding = color_encoding;
+  linear_color_encoding.Tf().SetTransferFunction(TransferFunction::kLinear);
```
Comment (sboukortt) [inline lib/jxl/dec_xyb.cc:255]: Is this missing the call to `CreateICC()`?

File: lib/jxl/decode.cc [hunk 4]
```diff
@@ -2653,28 +2662,63 @@ JxlDecoderStatus JxlDecoderGetFrameName(const JxlDecoder* dec, char* name,
 
 JxlDecoderStatus JxlDecoderSetPreferredColorProfile(
     JxlDecoder* dec, const JxlColorEncoding* color_encoding) {
+  return JxlDecoderSetOutputColorProfile(dec, color_encoding, nullptr, 0);
```
Comment (sboukortt) [inline lib/jxl/decode.cc:2665]: Parameter comments for the last two arguments would be helpful.

File: lib/jxl/decode_test.cc [hunk 5]
```diff
@@ -1864,12 +1895,41 @@ TEST(DecodeTest, SetPreferredColorProfileTestFromGray) {
...
+}
+
+class DecodeAllEncodingsVariantsTest
+    : public ::testing::TestWithParam<
+          std::tuple<jxl::test::ColorEncodingDescriptor, bool, bool>> {};
+JXL_GTEST_INSTANTIATE_TEST_SUITE_P(
+    DecodeAllEncodingsVariantsTestInstantiation, DecodeAllEncodingsVariantsTest,
+    ::testing::Combine(::testing::ValuesIn(jxl::test::AllEncodings()),
+                       ::testing::Values(false, true),
+                       ::testing::Values(false, true)),
```
Comment (sboukortt) [inline lib/jxl/decode_test.cc:1926]:
```suggestion
                       ::testing::Bool(),
                       ::testing::Bool()),
```

File: lib/jxl/dec_xyb.cc (hunk 3 above)
Comment (mo271) [reply lib/jxl/dec_xyb.cc:255]: This call moved to here:

https://github.com/libjxl/libjxl/blob/6cc2150ecb9f3b219fc18bec592b6dacf34bc071/lib/jxl/dec_cache.cc#L242

because it was causing failures in the case where we don't set a CMS

File: lib/jxl/dec_cache.cc (hunk 2 above)
Comment (mo271) [reply lib/jxl/dec_cache.cc:211]: done

File: lib/jxl/decode.cc (hunk 4 above)
Comment (mo271) [reply lib/jxl/decode.cc]: done

File: lib/include/jxl/decode.h [hunk 6]
```diff
@@ -877,7 +877,8 @@ JXL_EXPORT JxlDecoderStatus JxlDecoderSetOutputColorProfile(
  * @param cms structure representing a CMS implementation. See @ref
  * JxlCmsInterface for more details.
  */
-JXL_EXPORT void JxlDecoderSetCms(JxlDecoder* dec, JxlCmsInterface cms);
```
Comment (veluca93) [inline lib/include/jxl/decode.h:880]: What's the reason for this change?

What are the lifetime requirements for the cms parameter?

File: lib/jxl/dec_cache.cc [hunk 7]
```diff
@@ -195,23 +196,57 @@ Status PassesDecoderState::PreparePipeline(ImageBundle* decoded,
       }
     }
 
+    ColorEncoding linear_color_encoding;
+    if (frame_header.color_transform == ColorTransform::kXYB) {
+      linear_color_encoding = output_encoding_info.linear_color_encoding;
+    } else {
+      linear_color_encoding = ColorEncoding::LinearSRGB(
```
Comment (veluca93) [inline lib/jxl/dec_cache.cc:203]: I admit I don't exactly remember the semantics of linear_color_encoding, but is that supposed to change primaries wrt the original ones? I'd guess not...

File: lib/jxl/dec_cache.cc [hunk 8]
```diff
@@ -195,23 +196,57 @@ Status PassesDecoderState::PreparePipeline(ImageBundle* decoded,
...
-        builder.AddStage(std::move(to_linear_stage));
         linear = true;
       }
       builder.AddStage(std::move(tone_mapping_stage));
     }
 
     if (linear) {
-      builder.AddStage(GetFromLinearStage(output_encoding_info));
+      const size_t channels_src =
+          (output_encoding_info.linear_color_encoding.IsCMYK()
```
Comment (veluca93) [inline lib/jxl/dec_cache.cc:229]: ... For that matter, what's the linear_color_encoding supposed to be in the CMYK case?

File: lib/jxl/dec_cache.cc [hunk 9]
```diff
@@ -195,23 +196,57 @@ Status PassesDecoderState::PreparePipeline(ImageBundle* decoded,
...
+               ? 4
+               : output_encoding_info.linear_color_encoding.Channels());
+      const size_t channels_dst =
+          output_encoding_info.color_encoding.Channels();
+      const bool not_mixing_color_and_grey =
+          (channels_src == channels_dst ||
+           (channels_src == 4 && channels_dst == 3));
+      if (output_encoding_info.color_encoding_is_original ||
```
Comment (veluca93) [inline lib/jxl/dec_cache.cc:237]: A comment that explains what is happening with this if could be useful

File: lib/jxl/decode.cc [hunk 10]
```diff
@@ -2653,28 +2665,64 @@ JxlDecoderStatus JxlDecoderGetFrameName(const JxlDecoder* dec, char* name,
...
     return JXL_API_ERROR("too late to set the color encoding");
   }
-  if (dec->image_metadata.color_encoding.IsGray() &&
-      color_encoding->color_space != JXL_COLOR_SPACE_GRAY &&
-      dec->image_out_buffer_set && dec->image_out_format.num_channels < 3) {
-    return JXL_API_ERROR("Number of channels is too low for color output");
-  }
-  if (color_encoding->color_space == JXL_COLOR_SPACE_UNKNOWN) {
-    return JXL_API_ERROR("Unknown output colorspace");
+  // TODO(firsching): handle the non-icc case
```
Comment (veluca93) [inline lib/jxl/decode.cc:2687]: Do we want to do this before tagging a release?

File: lib/jxl/decode_test.cc [hunk 11]
```diff
@@ -1782,9 +1788,12 @@ void SetPreferredColorProfileTest(
       jxl::CreateTestJXLCodestream(jxl::Bytes(pixels.data(), pixels.size()),
                                    xsize, ysize, num_channels, params);
   auto all_encodings = jxl::test::AllEncodings();
-  all_encodings.push_back(
-      {jxl::ColorSpace::kXYB, jxl::WhitePoint::kD65, jxl::Primaries::kCustom,
-       jxl::TransferFunction::kUnknown, jxl::RenderingIntent::kPerceptual});
+  // TODO(firsching): understand why XYB does not work together with icc_dst.
```
Comment (veluca93) [inline lib/jxl/decode_test.cc:1791]: That'd be nice to understand:)

File: lib/jxl/decode_test.cc [hunk 12]
```diff
@@ -1864,12 +1896,40 @@ TEST(DecodeTest, SetPreferredColorProfileTestFromGray) {
...
+JXL_GTEST_INSTANTIATE_TEST_SUITE_P(
+    DecodeAllEncodingsVariantsTestInstantiation, DecodeAllEncodingsVariantsTest,
+    ::testing::Combine(::testing::ValuesIn(jxl::test::AllEncodings()),
+                       ::testing::Bool(), ::testing::Bool()),
+    DecodeAllEncodingsVariantsTestName);
+TEST_P(DecodeAllEncodingsVariantsTest, SetPreferredColorProfileTest) {
+  const auto& from = std::get<0>(GetParam());
+  bool icc_dst = std::get<1>(GetParam());
+  bool use_cms = std::get<2>(GetParam());
+  SetPreferredColorProfileTest(from, icc_dst, use_cms);
```
Comment (veluca93) [inline lib/jxl/decode_test.cc:1932]: Do we have a test that checks directly that asking for the same output in a different space gives the same image?

File: lib/include/jxl/decode.h (hunk 6 above)
Comment (mo271) [reply lib/include/jxl/decode.h:880]: I think this was due to some problem when cms-refactoring. 
Perhaps let's change it back, so we don't have to worry about the lifetime of `cms`. 
Plan to do something like this:
https://github.com/mo271/libjxl/compare/6a61ca734d29479b23aafeb208f6968ddc03988c..33984446603d920a9350f60d82bc91f65fb19010

File: lib/jxl/decode_test.cc (hunk 12 above)
Comment (mo271) [reply lib/jxl/decode_test.cc:1932]: No, not yet, indirectly yes, because the test image we use is always the same.
Should we?

File: lib/include/jxl/decode.h (hunk 6 above)
Comment (mo271) [reply lib/include/jxl/decode.h:880]: reverted

File: lib/jxl/decode.cc (hunk 10 above)
Comment (mo271) [reply lib/jxl/decode.cc]: I think we just forgot to delete this TODO and the non-icc is actually handled in the lines below.

File: lib/jxl/dec_cache.cc (hunk 8 above)
Comment (mo271) [reply lib/jxl/dec_cache.cc:229]: no idea. @sboukortt ?

File: lib/jxl/dec_cache.cc (hunk 9 above)
Comment (mo271) [reply lib/jxl/dec_cache.cc:237]: I added a comment.

File: lib/jxl/decode_test.cc (hunk 11 above)
Comment (mo271) [reply lib/jxl/decode_test.cc:1789]: https://github.com/libjxl/libjxl/issues/2946

File: lib/jxl/dec_cache.cc (hunk 8 above)
Comment (mo271) [reply lib/jxl/dec_cache.cc]: The `linear_color_encoding` is whatever is output from the from_xyb stage, also in the CMYK stage.
I added a comment to the new field `linear_color_encoding` to explain what it is.

File: lib/jxl/dec_cache.cc (hunk 7 above)
Comment (veluca93) [reply lib/jxl/dec_cache.cc:203]: Yes, my point is that that linear colorspace will likely not have SRGB primaries if the data encoding did not have SRGB primaries...

File: lib/jxl/dec_cache.cc [hunk 13]
```diff
@@ -195,23 +196,56 @@ Status PassesDecoderState::PreparePipeline(ImageBundle* decoded,
...
+          (output_encoding_info.orig_color_encoding.IsCMYK()
+               ? 4
+               : output_encoding_info.orig_color_encoding.Channels());
+      const size_t channels_dst =
+          output_encoding_info.color_encoding.Channels();
+      bool mixing_color_and_grey = (channels_dst != channels_src);
+      if ((output_encoding_info.color_encoding_is_original) ||
+          (!output_encoding_info.cms_set) || mixing_color_and_grey) {
+        // in those cases we only need a linear stage in other cases we attempt
+        // to obtain an cms stage
```
Comment (veluca93) [inline lib/jxl/dec_cache.cc:238]: This comment did not clarify to me why this condition is correct :)

Comment (mo271) [reply lib/jxl/dec_cache.cc]: I expanded the comment. 
The last condition was added when the `jxl_test` `RoundtripRGBToGray` was failing. 
If we don't have the first condition (`color_encoding_is_original`) everything should still work and indeed I just tried and all tests pass, but there is an superfluous cms_stage that is the identity.

Comment (veluca93) [reply lib/jxl/dec_cache.cc]: Ah, I see. Perhaps there should be a todo about the Gray condition then...

File: lib/jxl/dec_cache.cc (hunk 7 above)
Comment (mo271) [reply lib/jxl/dec_cache.cc:203]: Is that true, @sboukortt? Does it have the primaries of the data encoding or always SRGB primaries?

Comment (veluca93) [reply lib/jxl/dec_cache.cc:203]: After some consideration, we figured out (probably...) that we don't need that variable anymore.

**Review by veluca93 — DISMISSED** [review]

File: lib/jxl/dec_cache.cc (hunk 7 above)
Comment (mo271) [reply lib/jxl/dec_cache.cc]: Yes, removed it. For completeness: `linear_color_encoding` should be in the primaries of the ouput color space.

File: lib/jxl/dec_cache.cc (hunk 13 above)
Comment (mo271) [reply lib/jxl/dec_cache.cc]: added TODO

**Review by veluca93 — APPROVED** [review]
